### PR TITLE
feat: multiplayer undo / redo

### PR DESCRIPTION
**Pull request by Mrazator**
**TLDR;** Added support for multiplayer undo/redo, by calculating invertible increments and storing them inside the local-only undo/redo stacks.

* #5375

Other smaller changes include:

* Disabled footer buttons, when undo/redo stack is empty or zoom in/out reaches min/max value
* Added support for undo when collab is started ~ no cleanup and inability to undo
* Added recording on de-selection, while improving overall undo/redo of selections
* Fixed bug with colour picker not being recorded
* Fixed bugs with editing linear element editor not being recorded 
* Fixed bugs with editing group not being recorded
* Set potential stepping stones for
  * #2343
  * #3537
  * #3210

---

For more details check the high-level description of the introduced and modified components below.

image

Scene: https://link.excalidraw.com/readonly/jaraG0I63Qj6pPxqHVGn

**History** (_modified_)
* Saves emitted changes by the store into the local-only undo and redo stacks.
* Iterates history stack on undo/redo, when the recorded changes have no visible impact  - i.e. when elements/appState changes relate to deleted elements or a remote client modified it in the meantime to the same resulting state-specific cases covered in [here](https://github.com/excalidraw/excalidraw/pull/7348/files#diff-eac7c466d36d76a9f32c359f28f5795b88067eee030174e647794b8a9b967eaeR1378).
* Updates redo entry with the latest conflicting properties when applying undo entry (and vice versa), for details check [this](https://github.com/excalidraw/excalidraw/pull/7348/files#diff-eac7c466d36d76a9f32c359f28f5795b88067eee030174e647794b8a9b967eaeR549).

**Store** (_introduced_)
* Represents a component which abstracts away management of incremental changes.
* Captures all committed actions previously recorded by the history into an internal always up-to-date snapshot.
  * Remote actions filter out yet uncommitted elements (as a result of local async action, such as dragging, editing, resizing etc.) and only update the snapshot, without calculating the increment.
* Calculates and emits the store increment as a set of changes based on the previously captured snapshot.
* Exposes event listener for other components interested in the store increments - for now only history is tied to it, but in the future, we could also tie other components, such as collab, storage or versioning.

**Change** (_introduced_)
* Represents a point in the state history as a result of a user action (event).
* Encapsulates a payload in the form of delta/s and contains methods for calculating/applying/inverting the change.
* Contains additional logic which is specific to each change implementation, i.e. conflict resolution logic for `ElementsChange`, which mostly relates to the text container bindings in the multiplayer scenarios (specific cases covered in [here](https://github.com/excalidraw/excalidraw/pull/7348/files#diff-eac7c466d36d76a9f32c359f28f5795b88067eee030174e647794b8a9b967eaeR751).
* For now, two specific implementations were introduced, `AppStateChange` (purely local) and `ElementsChange` (local, but could conflict with remote changes).
  * Potentially, each conflicting change could hold additional metadata, such as `id`, `type`, `source`, and logical `timestamp`, which would extend it with CRDT-like characteristics.

**Delta** (_introduced_)
* Represents a delta between two objects on the properties level in the form of a pure object (side-effects free).
* Encapsulates `deleted` and `inserted` partials, where `deleted` is a set of all the previous (removed) values and `inserted` is a set of all the next (added, updated) values.
* Can be stored inside the history stacks, inversed when applying undo/redo or used for time travel purposes.
* Inverse operation simply replaces `deleted` with `inserted` and vice versa - as visible [here](https://github.com/excalidraw/excalidraw/pull/7348/files#diff-8f92fc99a0f79360fa765f124bc4c4a660e656192d22e80118fb311b5e4d7ca1R289).
 * To sneak peek the `Delta` format take a look at the following [gist](https://gist.github.com/Mrazator/56b6d039bba6cfe890f317be2f26621a).

### CODE REVIEW

File: packages/excalidraw/fractionalIndex.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+    if (!element.fractionalIndex) {
+      return false;
+    }
+
+    const successor: ExcalidrawElement | undefined = elements[i + 1];
+
+    if (
+      successor?.fractionalIndex &&
+      element.fractionalIndex > successor.fractionalIndex
```
Comment (Mrazator) [inline packages/excalidraw/fractionalIndex.ts:89]: @Mrazator here should be `>=` here instead, as we would like to catch also equalities in the future. Potentially could be also unified with `isValidFractionalIndex` below.

File: packages/excalidraw/change.ts [hunk 2]
```diff
@@ -0,0 +1,1262 @@
...
+
+    for (const { container, boundText } of boxesToRedraw.values()) {
+      if (container.isDeleted || boundText.isDeleted) {
+        // Skip on deleted container or bound text, as it doesn't result in visible changes so we don't need to redraw
+        continue;
+      }
+
+      // TODO: this is a huge bottleneck which can take up to 96% of undo / redo computation time
+      // TODO: refactor mutations away, so we couln't end up in an incosistent state
+      redrawTextBoundingBox(boundText, container, false);
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts:1164]: cc @dwelle @ad1992

File: excalidraw-app/collab/reconciliation.ts [hunk 3]
```diff
@@ -21,7 +19,7 @@ const shouldDiscardRemoteElement = (
     // local element is being edited
     (local.id === localAppState.editingElement?.id ||
       local.id === localAppState.resizingElement?.id ||
-      local.id === localAppState.draggingElement?.id ||
+      local.id === localAppState.draggingElement?.id || // Is this still valid? As draggingElement is selection element, which is never part of the elements array
```
Comment (Mrazator) [inline excalidraw-app/collab/reconciliation.ts:22]: @dwelle seemed weird when debugging, check the comment

File: packages/excalidraw/components/App.tsx [hunk 4]
```diff
@@ -3524,18 +3563,57 @@ class App extends React.Component<AppProps, AppState> {
...
+           *
+           * This is also a breaking change for all local `updateScene` calls without set `commitToStore` to true,
+           * as it makes such updates impossible to undo (previously they were undone coincidentally with the switch to the whole previously captured snapshot by history).
+           *
+           * WARN: be careful here as moving it elsewhere coulb break the history for remote client without noticing
+           * - we need to find a way to test two concurrent client updates simultaneously, while having access to both stores & histories.
+           */
+          this.store.shouldUpdateSnapshot();
+
+          nextCommittedElements = this.store.ignoreUncomittedElements(
```
Comment (Mrazator) [inline packages/excalidraw/components/App.tsx:3600]: As written in the description of `ignoreUncomittedElements`, once we will be exchanging just increments and ephemerals (i.e. yet uncomitted actions) won't be recorded, this filtering won't be necessary.

File: packages/excalidraw/components/App.tsx [hunk 5]
```diff
@@ -8735,13 +8822,21 @@ class App extends React.Component<AppProps, AppState> {
...
         fileHandle,
       );
       if (ret.type === MIME_TYPES.excalidraw) {
+        // Restore the fractional indices by mutating elements and update the
+        // store snapshot, otherwise we would end up with duplicate indices
+        restoreFractionalIndices(elements.concat(ret.data.elements));
+        this.store.snapshot = this.store.snapshot.clone(
+          arrayToMap(elements),
+          this.state,
+        );
```
Comment (Mrazator) [inline packages/excalidraw/components/App.tsx:8839]: Above updates all fractional indices in the snapshot as we purposefully never delete from it even though the elements were removed from the scene (so we could calculate the deltas and always go back). Not updating the indices could lead to us having duplicate indices in the scene, after a series of undo / redo commands. For details check `scene import via drag&drop should create new history entry`.

File: packages/excalidraw/change.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts]: The bottleneck is gone now due to #7693

File: packages/excalidraw/components/App.tsx (hunk 5 above)
Comment (Mrazator) [reply packages/excalidraw/components/App.tsx:9061]: Update, the above sync updates only invalid indices (likely in the dropped scene) due to improvements in #7693

File: packages/excalidraw/frame.ts [hunk 6]
```diff
@@ -29,7 +29,7 @@ import { ReadonlySetLike } from "./utility-types";
 
 // --------------------------- Frame State ------------------------------------
 export const bindElementsToFramesAfterDuplication = (
-  nextElements: ExcalidrawElement[],
+  nextElements: readonly ExcalidrawElement[],
```
Comment (Mrazator) [inline packages/excalidraw/frame.ts:32]: double-check why

File: packages/excalidraw/tests/data/restore.test.ts [hunk 7]
```diff
@@ -290,12 +312,11 @@ describe("restoreElements", () => {
     expect(restoredElements).toEqual([
       expect.objectContaining({
         id: rectangle.id,
-        version: rectangle_modified.version + 1,
+        version: rectangle_modified.version + 2,
```
Comment (Mrazator) [inline packages/excalidraw/tests/data/restore.test.ts:315]: not particularly in favour of the double version bump, but I expect we could live some time without it;

ideally it should be solved by generically batching mutations (with possible rollback)

File: packages/excalidraw/tests/scene/export.test.ts [hunk 8]
```diff
@@ -15,8 +15,18 @@ describe("exportToSvg", () => {
   const ELEMENT_HEIGHT = 100;
   const ELEMENT_WIDTH = 100;
   const ELEMENTS = [
-    { ...diamondFixture, height: ELEMENT_HEIGHT, width: ELEMENT_WIDTH },
-    { ...ellipseFixture, height: ELEMENT_HEIGHT, width: ELEMENT_WIDTH },
+    {
+      ...diamondFixture,
+      height: ELEMENT_HEIGHT,
+      width: ELEMENT_WIDTH,
+      index: "A0",
```
Comment (Mrazator) [inline packages/excalidraw/tests/scene/export.test.ts:22]: change to 'a0' , respectively 'a1' below

Comment (Mrazator) [reply packages/excalidraw/tests/scene/export.test.ts:22]: #7359

File: excalidraw-app/collab/reconciliation.ts [hunk 9]
```diff
@@ -35,120 +38,48 @@ const shouldDiscardRemoteElement = (
...
-    arrayToMapWithIndex<ExcalidrawElement>(localElements);
-
-  const reconciledElements: ExcalidrawElement[] = localElements.slice();
-
-  const duplicates = new WeakMap<ExcalidrawElement, true>();
-
-  let cursor = 0;
-  let offset = 0;
+): ReconciledExcalidrawElement[] => {
+  const localElementsData = arrayToMap(localElements);
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply excalidraw-app/collab/reconciliation.ts:47]: #7359

File: packages/excalidraw/history.ts [hunk 10]
```diff
@@ -1,265 +1,195 @@
...
-          // don't store multi-point arrow if still has only one point
-          if (
-            appState.multiElement &&
-            appState.multiElement.id === element.id &&
-            element.points.length < 2
-          ) {
-            return elements;
-          }
-
-          elements.push({
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/history.ts:122]: - one undo could cause an infinite loop, as undo leads to another `record` being triggered, pushing same increment into the undo stack, due to:
```
    if (
      this.state.editingLinearElement &&
      !this.state.selectedElementIds[this.state.editingLinearElement.elementId]
    ) {
      // defer so that the storeAction flag isn't reset via current update
      setTimeout(() => {
        // execute only if the condition still holds when the deferred callback
        // executes (it can be scheduled multiple times depending on how
        // many times the component renders)
        this.state.editingLinearElement &&
          this.actionManager.executeAction(actionFinalize);
      });
    }
```
- related, undo /redo should open again the linear element editor
- related, force deleted element's recovered through snapshot should set zIndex & visibleChange flags separately

File: packages/excalidraw/change.ts [hunk 11]
```diff
@@ -0,0 +1,1273 @@
...
+      (element: OrderedExcalidrawElement) => (partial: ElementPartial) => {
+        const latestPartial: { [key: string]: unknown } = {};
+
+        for (const key of Object.keys(partial) as Array<keyof typeof partial>) {
+          if (
+            key === "boundElements" ||
+            key === "groupIds" ||
+            key === "customData"
+          ) {
+            // It doesn't make sense to update the above props since:
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts:762]: Check the implications to `customData`

File: packages/excalidraw/history.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/history.ts:122]: fixed

File: packages/excalidraw/change.ts [hunk 12]
```diff
@@ -0,0 +1,1342 @@
...
+  assertNever,
+  isShallowEqual,
+} from "./utils";
+
+/**
+ * Represents the difference between two objects of the same type.
+ *
+ * Both `deleted` and `inserted` partials represent the same set of added, removed or updated properties, where:
+ * - `deleted` is a set of all the previous (removed) values
+ * - `inserted` is a set of all the next (added, updated) values
```
Comment (Mrazator) [inline packages/excalidraw/change.ts:43]: Either emphasise it's not always (for reference values only if we postprocess them), or make it generic-enough to be so that all the reference values are always postprocessed automatically.

**Comment by Mrazator** [comment]
Test edge case related to force deleting invisibly small elements.

File: packages/excalidraw/actions/actionFinalize.tsx [hunk 13]
```diff
@@ -93,7 +95,11 @@ export const actionFinalize = register({
           });
         }
       }
+
       if (isInvisiblySmallElement(multiPointElement)) {
+        // TODO: not a great idea here as it could be recorded by the store,
+        // so the invisible element could be restored by the undo/redo, which is not what we want
```
Comment (Mrazator) [inline packages/excalidraw/actions/actionFinalize.tsx:101]: Test the edge cases related to deleting invisibly small elements.

File: packages/excalidraw/change.ts [hunk 14]
```diff
@@ -0,0 +1,1342 @@
...
+    const nextAppState = {
+      ...appState,
+      ...directlyApplicablePartial,
+      selectedElementIds: mergedSelectedElementIds,
+      selectedGroupIds: mergedSelectedGroupIds,
+      selectedLinearElement:
+        typeof selectedLinearElementId !== "undefined"
+          ? selectedLinearElement // element was either inserted or deleted
+          : appState.selectedLinearElement, // otherwise assign what we had before
+      editingLinearElement:
```
Comment (Mrazator) [inline packages/excalidraw/change.ts:356]: Cover with test cases

File: packages/excalidraw/history.ts [hunk 15]
```diff
@@ -1,265 +1,195 @@
...
+    const [nextElements, elementsContainVisibleChange] =
+      this.elementsChange.applyTo(elements, snapshot.elements);
+
+    const [nextAppState, appStateContainsVisibleChange] =
+      this.appStateChange.applyTo(appState, nextElements);
+
+    const appliedVisibleChanges =
+      elementsContainVisibleChange || appStateContainsVisibleChange;
+
+    return [nextElements, nextAppState, appliedVisibleChanges];
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/history.ts:177]: In worst case refresh restore us back into a consistent state?

File: packages/excalidraw/change.ts [hunk 16]
```diff
@@ -0,0 +1,1273 @@
...
+  /**
+   * It is necessary to post process the partials in case of reference values,
+   * for which we need to calculate the real diff between `deleted` and `inserted`.
+   */
+  private static postProcess(
+    deleted: ElementPartial,
+    inserted: ElementPartial,
+  ): [ElementPartial, ElementPartial] {
+    if (deleted.boundElements && inserted.boundElements) {
+      const deletedDifferences = arrayToObject(
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts:1280]: I rechecked reference props of other elements and it seems that we don't need have do this kind of merging (not to override other clients changes) in other cases, at least not now.

Closest next thing might be merging of linear element points - in case multiple clients modify the same elements and later undo / redo is perfomed. ATM one client will completely override the other clients changes, but it's even questionable if we should perform merge in such case as it might lead to broken linear element when we would merge the changes.

In the future, we might consider a special runtime brand or a decorator, which we would put over a given instance and it would automatically postprocess it (getting just the changes) & automatically merge the changes when applying it.

File: packages/excalidraw/change.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts:806]: As stated in https://github.com/excalidraw/excalidraw/pull/7348#discussion_r1521718445, we might offer a custom decorator / brand to merge extract just the changes of reference types, such as `customData` (perhaps based on some id) and merge only changes. However, right now it does not seem to be needed, so let's not over-engineer. 

For now basic shallow compare will be performed on the first level of `customData` props, if a difference will be found, whole `customData` will be recorded as part of the increment, in the state before (`deleted`) and after (`inserted`).

File: packages/excalidraw/change.ts [hunk 17]
```diff
@@ -0,0 +1,1273 @@
...
+        case "editingLinearElement":
+          nextAppState[key] = AppStateChange.filterLinearElement(
+            nextAppState[key],
+            nextElements,
+            visibleDifferenceFlag,
+          );
+          break;
+        case "editingGroupId":
+        case "selectedGroupIds":
+          // Currently we don't have an index of elements by groupIds, which means that
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts:496]: Updated the comment, not worth doing ATM, as worst scenario is an empty undo, which likely won't happen very often.

File: excalidraw-app/collab/reconciliation.ts (hunk 3 above)
Comment (Mrazator) [reply excalidraw-app/collab/reconciliation.ts:27]: Added TODO to check in the future

File: packages/excalidraw/change.ts [hunk 18]
```diff
@@ -0,0 +1,1273 @@
...
+    if (prevObject === nextObject) {
+      return Delta.empty();
+    }
+
+    const deleted = {} as Partial<T>;
+    const inserted = {} as Partial<T>;
+
+    // O(n^3) here, but it's not as bad as it looks:
+    // - we do this only on history recordings, not on every frame
+    // - we do this only on changed elements
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts]: Based on manual testing, both snapshot perf. & applyTo seems great, with negligible implications.

image
.

File: packages/excalidraw/change.ts (hunk 12 above)
Comment (Mrazator) [reply packages/excalidraw/change.ts:43]: Won't be generic for now, reasoning https://github.com/excalidraw/excalidraw/pull/7348#discussion_r1521718445

Comment & description should be updates

File: packages/excalidraw/change.ts [hunk 19]
```diff
@@ -0,0 +1,1273 @@
...
+      const {
+        boundElements: addedBoundElements,
+        groupIds: addedGroupIds,
+        ...directlyApplicablePartial
+      } = delta.inserted;
+
+      const { boundElements, groupIds } = element;
+
+      let nextBoundElements = boundElements;
+      if (addedBoundElements?.length || removedBoundElements?.length) {
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts:981]: Actually bound elements cannot be generic, due to keeping both bound text elements and arrows as part of `boundElements` array - hence the necessity to unbind existing text bindings, once a new text binding is added as part of the increment.

We should likely separate bound text into a standalone property, so it does not collide. It also sorts of reflects parent-child relationship, as opposed to arrow bindings.

Related https://github.com/excalidraw/excalidraw/pull/7348#discussion_r1521718445

File: packages/excalidraw/change.ts [hunk 20]
```diff
@@ -0,0 +1,1273 @@
...
+    const applyDelta = ElementsChange.createApplier(elements, flags);
+
+    const changes = new Map<string, OrderedExcalidrawElement>();
+
+    function setElements(
+      ...changedElements: (OrderedExcalidrawElement | undefined)[]
+    ) {
+      for (const element of changedElements) {
+        if (element) {
+          elements.set(element.id, element);
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts:866]: Could be an issue for syncing, let's check again. The biggest concern is likely that related text binding fixes would have to work on top of partials as well.

File: packages/excalidraw/actions/actionHistory.tsx [hunk 21]
```diff
@@ -1,105 +1,125 @@
...
+      elements: nextElements,
+      storeAction: StoreAction.UPDATE,
     };
   }
-  return { commitToHistory };
+
+  return { storeAction: StoreAction.NONE };
+};
+
+const useEmitter = (emitter: History["onHistoryChangeEmitter"]) => {
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/actions/actionHistory.tsx]: Done

File: packages/excalidraw/fractionalIndex.ts (hunk 1 above)
Comment (Mrazator) [reply packages/excalidraw/fractionalIndex.ts]: Solved differently as part of #7359, we make sure indices are always deduplicated.

File: packages/excalidraw/change.ts [hunk 22]
```diff
@@ -0,0 +1,1343 @@
...
+
+    const mergedSelectedGroupIds = Delta.merge(
+      appState.selectedGroupIds,
+      addedSelectedGroupIds,
+      removedSelectedGroupIds,
+    );
+
+    const selectedLinearElement =
+      selectedLinearElementId && elements.has(selectedLinearElementId)
+        ? new LinearElementEditor(
```
Comment (Mrazator) [inline packages/excalidraw/change.ts:331]: Double-check if it makes sense keeping whole `LinearElementEditor instance state as part the `appState`, as from the history point of view it does not.

File: packages/excalidraw/change.ts [hunk 23]
```diff
@@ -0,0 +1,1374 @@
...
+
+      if (existingElement) {
+        const addedElement = applyDelta(existingElement, delta);
+        changed.set(addedElement.id, addedElement);
+        added.set(addedElement.id, addedElement);
+      }
+    }
+
+    // Performs mutations, but mostly on new instances of elements
+    // trying to avoid potential mutations of previous state leading to unexpected incosistencies
```
Comment (Mrazator) [inline packages/excalidraw/change.ts:905]: Could we fallback to previous state if we detect inconsistencies (assuming we could somehow)?

File: packages/excalidraw/change.ts [hunk 24]
```diff
@@ -0,0 +1,1374 @@
...
+        if (element) {
+          // As the element was brought from the snapshot, it automatically results in possible zindex difference
+          flags.containsZindexDifference = true;
+
+          // As the element was force deleted, we need to check if adding it back results in a visible change
+          if (
+            partial.isDeleted === false ||
+            (partial.isDeleted !== true && element.isDeleted === false)
+          ) {
+            flags.containsVisibleDifference = true;
```
Comment (Mrazator) [inline packages/excalidraw/change.ts:1026]: Test all these cases - restoring from snapshot for added/updated/removed, setting flags for zindex differences & visible differences.

File: packages/excalidraw/change.ts [hunk 25]
```diff
@@ -0,0 +1,1343 @@
...
+
+    for (const [id, delta] of this.removed.entries()) {
+      const existingElement = getElement(id, delta.inserted);
+
+      if (existingElement) {
+        const removedElement = applyDelta(existingElement, delta);
+
+        setElements(
+          removedElement,
+          ElementsChange.whenTextContainer(removedElement)?.removeBoundText(
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts]: Described

File: packages/excalidraw/tests/data/restore.test.ts (hunk 7 above)
Comment (Mrazator) [reply packages/excalidraw/tests/data/restore.test.ts:315]: Related to #7359

File: packages/excalidraw/change.ts (hunk 22 above)
Comment (Mrazator) [reply packages/excalidraw/change.ts:334]: Added TODO

File: packages/excalidraw/change.ts [hunk 26]
```diff
@@ -0,0 +1,1273 @@
...
+    private readonly removed: Map<string, Delta<ElementPartial>>,
+    private readonly updated: Map<string, Delta<ElementPartial>>,
+  ) {}
+
+  public static create(
+    added: Map<string, Delta<ElementPartial>>,
+    removed: Map<string, Delta<ElementPartial>>,
+    updated: Map<string, Delta<ElementPartial>>,
+  ) {
+    if (import.meta.env.DEV || import.meta.env.MODE === ENV.TEST) {
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts:618]: Would't be great for perf in prod, so let's not spam the sentry for now.

File: packages/excalidraw/change.ts (hunk 23 above)
Comment (Mrazator) [reply packages/excalidraw/change.ts:905]: For now we will rely on restore, in the future we could persist the history stack so one could always undo/redo (even after refresh).

File: packages/excalidraw/change.ts [hunk 27]
```diff
@@ -0,0 +1,1273 @@
...
+  ): ElementsChange {
+    if (prevElements === nextElements) {
+      return ElementsChange.empty();
+    }
+
+    const added = new Map<string, Delta<ElementPartial>>();
+    const removed = new Map<string, Delta<ElementPartial>>();
+    const updated = new Map<string, Delta<ElementPartial>>();
+
+    // This might be needed only in same edge cases, like during collab, when `isDeleted` elements get removed
```
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts:681]: There might be a lot of intended deletions, so we would just spam the sentry for now.

File: packages/excalidraw/change.ts (hunk 24 above)
Comment (Mrazator) [reply packages/excalidraw/change.ts:1026]: Tested as part of the scene drag&drop

File: packages/excalidraw/change.ts [hunk 28]
```diff
@@ -0,0 +1,1394 @@
...
+    );
+
+    return nextBoundTextElements.length ? nextBoundTextElements : null;
+  }
+
+  /**
+   * Fixing text bindings so we won't end up in an incosistent state after applying changes.
+   *
+   * The logic is similar to one `fixBindingsAfterDeletion`, `repairBoundElement` and `repairContainerElement`, but serves a different purpose.
+   * While the above functions are designed to ensure consistent state during restore, these are designed to resolve conflicts in the least intrusive way during concurrent updates.
```
Comment (Mrazator) [inline packages/excalidraw/change.ts:1105]: Test once again current bindings behaviour and write down pros & cons, as i.e.:
- arrows will remain bound after element deletion through history, which could be confusing when trying rebind again
- elements will remain bound to frames after frame deletion through history, which could be confusing when trying to rebind to a different frame

File: packages/excalidraw/change.ts [hunk 29]
```diff
@@ -0,0 +1,1394 @@
...
+    return (
+      this.added.size === 0 &&
+      this.removed.size === 0 &&
+      this.updated.size === 0
+    );
+  }
+
+  /**
+   * Update only `updated` delta/s based on the existing elements.
+   * In comparison, `added` and `removed` deltas are not updated as it would cause our invariants to break.
```
Comment (Mrazator) [inline packages/excalidraw/change.ts:794]: Actually we should update added & removed, but not updating `isDeleted` flag, so the invariants will stand.
+ add test cases

Comment (Mrazator) [reply packages/excalidraw/change.ts]: Done

File: packages/excalidraw/change.ts (hunk 28 above)
Comment (Mrazator) [reply packages/excalidraw/change.ts:1105]: Likely does not make sense leaving it as is & will need to be fixed

File: packages/excalidraw/change.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (Mrazator) [reply packages/excalidraw/change.ts]: In the end we did it already.

File: packages/excalidraw/change.ts (hunk 28 above)
Comment (Mrazator) [reply packages/excalidraw/change.ts]: Fixed

File: packages/excalidraw/change.ts (hunk 14 above)
Comment (Mrazator) [reply packages/excalidraw/change.ts]: Done

File: packages/excalidraw/utils.ts [hunk 30]
```diff
@@ -671,6 +671,18 @@ export const arrayToMapWithIndex = <T extends { id: string }>(
...
+ */
+export const arrayToObject = <T>(
+  array: readonly T[],
+  groupBy?: (value: T) => string,
+) =>
+  array.reduce((acc, value) => {
+    acc[groupBy ? groupBy(value) : String(value)] = value;
+    return acc;
+  }, {} as { [key: string]: T });
+
```
Comment (dwelle) [inline packages/excalidraw/utils.ts:685]: I'd consider switching to a Map here as [allegedly Maps are more optimized for additions/deletions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) (though we'd have to benchmark).

Comment (Mrazator) [reply packages/excalidraw/utils.ts:685]: Considering it is scarcely used (currently on `boundElements` only), it would not likely result in any perf. benefits. Actually it might be even worse, since the idea behind `Delta` is that it works on plain array / object, which do not have to be instantiated on load (i.e. when restoring from storage, receiving from other clients, etc.). Also, if we would instantiate Map here for each array, we would have to then go back and forth between the Map and Object, since all the `Delta` utility methods, such as `getLeftDifferences`, `getRightDifferences` work on top of objects.

**Review by dwelle — APPROVED** [review]
🔥🚀

**Comment by satoren** [comment]
Are there any plans to release a version of the npm package with this fix?

**Comment by Mrazator** [comment]
Yes, likely in couple of weeks.
